pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,364 @@
+	watcherName                 = "L7Watcher"
+	resyncPeriod  time.Duration = 0
+	minRetryDelay               = 5 * time.Second
+	maxRetryDelay               = 300 * time.Second
+	// Default number of workers processing an Egress change.
+	defaultWorkers            = 4
+	l7VisibilityAnnotationKey = "l7Visibility.antrea.io"
+)
+
+type l7Watcher struct {
+	ofClient       openflow.Client
+	interfaceStore interfacestore.InterfaceStore
+
+	podInformer     cache.SharedIndexInformer
+	PodLister       corelisters.PodLister
+	podListerSynced cache.InformerSynced
+
+	namespaceInformer     cache.SharedIndexInformer
+	NamespaceLister       corelisters.NamespaceLister
+	namespaceListerSynced cache.InformerSynced
+
+	l7Reconciler      *Reconciler
+	l7podToTCMap      map[string]string
+	l7podToTCMapMutex sync.RWMutex
+	queue             workqueue.RateLimitingInterface
+	once              sync.Once
+}
+
+func NewL7Watcher(
+	ofClient openflow.Client,
+	interfaceStore interfacestore.InterfaceStore,
+	podInformer cache.SharedIndexInformer,
+	namespaceInformer coreinformers.NamespaceInformer) *l7Watcher {
+	l7Reconciler := NewReconciler()
+	l7podToTCMap := make(map[string]string)
+	l7w := &l7Watcher{
+		ofClient:              ofClient,
+		interfaceStore:        interfaceStore,
+		podInformer:           podInformer,
+		PodLister:             corelisters.NewPodLister(podInformer.GetIndexer()),
+		podListerSynced:       podInformer.HasSynced,
+		namespaceInformer:     namespaceInformer.Informer(),
+		NamespaceLister:       namespaceInformer.Lister(),
+		namespaceListerSynced: namespaceInformer.Informer().HasSynced,
+		l7Reconciler:          l7Reconciler,
+		l7podToTCMap:          l7podToTCMap,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "l7Watchergroup"),
+	}
+	l7w.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7w.addPod,
+			UpdateFunc: l7w.updatePod,
+			DeleteFunc: l7w.deletePod,
+		},
+		resyncPeriod,
+	)
+	l7w.namespaceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7w.addNamespace,
+			UpdateFunc: l7w.updateNamespace,
+			DeleteFunc: l7w.deleteNamespace,
+		},
+		resyncPeriod,
+	)
+	return l7w
+}
+
+func (l7w *l7Watcher) Run(stopCh <-chan struct{}) {
+	defer l7w.queue.ShutDown()
+	klog.InfoS("Starting", "Watcher", watcherName)
+	defer klog.InfoS("Shutting down", "Watcher", watcherName)
+
+	if !cache.WaitForNamedCacheSync(watcherName, stopCh, l7w.podListerSynced, l7w.namespaceListerSynced) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(l7w.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (l7w *l7Watcher) worker() {
+	for l7w.processNextWorkItem() {
+	}
+}
+
+func (l7w *l7Watcher) processNextWorkItem() bool {
+	obj, quit := l7w.queue.Get()
+	if quit {
+		return false
+	}
+	defer l7w.queue.Done(obj)
+
+	if podObj, ok := obj.(*v1.Pod); !ok {
```

Comment:
I didn't understand why using namespaced name over pod object will guarantee non parallel processing of the same obj. 
The other reason to use a pod object is to avoid the further api calls when looking for annotations.

Reply:
It's better to check other controllers' examples first if you are not familiar with controller pattern.
1. Please check workqueue doc https://pkg.go.dev/k8s.io/client-go/util/workqueue or code if you want to understand how it guarantees a single item will not be processed multiple times concurrently and why the key must be the name, instead of the object.
2. There is no extra API call, it's in-memory query when using PodLister.

Reply:
Thanks for the doc! code will be updated.